File: 148.Sort-List.go
// https://leetcode.com/problems/sort-list/
//
// algorithms
// Hard (33.79%)
// Total Accepted:    168,913
// Total Submissions: 499,908
// beats 100.0% of golang submissions

// 如果不允许直接修改 node.Val，那就先找到中间节点，再归并排序
package leetcode

import "sort"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
	var vals []int

	for tmp := head; tmp != nil; {
		vals = append(vals, tmp.Val)
		tmp = tmp.Next
	}

	sort.Sort(sort.IntSlice(vals))

	tmp := head
	for _, n := range vals {
		tmp.Val = n
		tmp = tmp.Next
	}

	return head
}

File: 148.Sort-List.java
// https://leetcode.com/problems/sort-list/
//
// algorithms
// Medium (34.81%)
// Total Accepted:    179,359
// Total Submissions: 515,226


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode l = head, r = head.next, pre = null;
        
        while (r != null) {
            pre = l;
            l = l.next;
            r = r.next;
            if (r != null) {
                r = r.next;
            }
        }
        
        pre.next = null;
        
        r = this.sortList(l);
        l = this.sortList(head);
        ListNode res = new ListNode(-1);
        ListNode tmp = res;
        
        while (l != null && r != null) {
            if (l.val <= r.val) {
                res.next = l;
                l = l.next;
            } else {
                res.next = r;
                r = r.next;
            }
            res = res.next;
            res.next = null;
        }
        
        if (l != null) {
            res.next = l;
        }
        if (r != null) {
            res.next = r;
        }
        
        return tmp.next;
    }
}